File: cpp/min_stack.cpp
class MinStack {
public:
    void push(int x) {
        if (minStack.empty() || x<=minStack.top())
            minStack.push(x);
        numStack.push(x);
    }

    void pop() {
        if (!numStack.empty() && !minStack.empty())
        {
            if (numStack.top() == minStack.top())
                minStack.pop();
            numStack.pop();
        }
    }

    int top() {
        return numStack.top();
    }

    int getMin() {
        return minStack.top();
    }
private:
    stack<int> numStack;
    stack<int> minStack;
};

File: java/min_stack.java
class MinStack {
    Stack<Integer> numStack = new Stack<Integer>();
    Stack<Integer> minStack = new Stack<Integer>();
    
    public void push(int x) {
        if (minStack.empty() || x<=minStack.peek())
            minStack.push(x);
        numStack.push(x);
    }

    public void pop() {
        if (!numStack.empty() && !minStack.empty())
        {
            if (numStack.peek().equals(minStack.peek()))
                minStack.pop();
            numStack.pop();
        }
    }

    public int top() {
            return numStack.peek();
    }

    public int getMin() {
            return minStack.peek();
    }
}